connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala

```diff
@@ -572,6 +726,101 @@ private class KeyValueGroupedDatasetImpl[K, V, IK, IV](
+        .addAllInitialGroupingExpressions(initialState.getGroupingExpressions)
+    }
+  }
+
+  override def mapGroupsWithState[S: Encoder, U: Encoder](timeoutConf: GroupStateTimeout)(
+    func: (K, Iterator[V], GroupState[S]) => U): Dataset[U] = {
+    val nf = if (valueMapFunc == UdfUtils.identical()) {
+      func
+    } else {
+      UdfUtils.mapValuesAdaptor(func, valueMapFunc)
+    }
+    val outputEncoder = encoderFor[U]
+    sparkSession.newDataset[U](outputEncoder) { builder =>
+      builder.getFlatMapGroupsWithStateBuilder
+        .setInput(plan.getRoot)
+        .addAllGroupingExpressions(groupingExprs)
+        .setFunc(getUdf(nf, outputEncoder)(ivEncoder))
+        .setIsMapGroupsWithState(true)
+        .setOutputMode(OutputMode.Update.toString)
+        .setTimeoutConf(timeoutConf.toString)
+    }
+  }
+
+  override def flatMapGroupsWithState[S: Encoder, U: Encoder](
+      outputMode: OutputMode,
+      timeoutConf: GroupStateTimeout,
+      initialState: KeyValueGroupedDataset[K, S])(
+      func: (K, Iterator[V], GroupState[S]) => Iterator[U]): Dataset[U] = {
+    if (outputMode != OutputMode.Append && outputMode != OutputMode.Update) {
+      throw new IllegalArgumentException("The output mode of function should be append or update")
+    }
+    val nf = if (valueMapFunc == UdfUtils.identical()) {
+      func
+    } else {
+      UdfUtils.mapValuesAdaptor(func, valueMapFunc)
+    }
+    val outputEncoder = encoderFor[U]
+    sparkSession.newDataset[U](outputEncoder) { builder =>
+      builder.getFlatMapGroupsWithStateBuilder
+        .setInput(plan.getRoot)
+        .addAllGroupingExpressions(groupingExprs)
+        .setFunc(getUdf(nf, outputEncoder)(ivEncoder))
+        .setIsMapGroupsWithState(false)
+        .setOutputMode(outputMode.toString)
+        .setTimeoutConf(timeoutConf.toString)
+        .addAllInitialGroupingExpressions(initialState.getGroupingExpressions)
+    }
+  }
+
+  override def flatMapGroupsWithState[S: Encoder, U: Encoder](
+      outputMode: OutputMode,
+      timeoutConf: GroupStateTimeout)(
+      func: (K, Iterator[V], GroupState[S]) => Iterator[U]): Dataset[U] = {
+    if (outputMode != OutputMode.Append && outputMode != OutputMode.Update) {
+      throw new IllegalArgumentException("The output mode of function should be append or update")
+    }
+    val nf = if (valueMapFunc == UdfUtils.identical()) {
+      func
+    } else {
+      UdfUtils.mapValuesAdaptor(func, valueMapFunc)
+    }
+    val outputEncoder = encoderFor[U]
+    sparkSession.newDataset[U](outputEncoder) { builder =>
+      builder.getFlatMapGroupsWithStateBuilder
+        .setInput(plan.getRoot)
+        .addAllGroupingExpressions(groupingExprs)
+        .setFunc(getUdf(nf, outputEncoder)(ivEncoder))
+        .setIsMapGroupsWithState(false)
+        .setOutputMode(outputMode.toString)
+        .setTimeoutConf(timeoutConf.toString)
+    }
+  }
+
+  override def getGroupingExpressions: java.util.List[proto.Expression] = {
```

Comment:
Can we follow the implementation here:
https://github.com/apache/spark/blob/db3f85b24498d4904134ed8f14b570fde6550060/connector/connect/client/jvm/src/main/scala/org/apache/spark/sql/KeyValueGroupedDataset.scala#L685-L703

To get rid of this function completely?

Reply:
Thanks for the pointer, yeah this is a better way to handle it!